src/core/src/op/read_value.cpp

```diff
@@ -50,36 +55,84 @@ op::v6::ReadValue::ReadValue(const Output<Node>& init_value, const shared_ptr<Va
 
 void op::v6::ReadValue::validate_and_infer_types() {
     OV_OP_SCOPE(v6_ReadValue_validate_and_infer_types);
-    const auto arg_t = get_input_element_type(0);
-    const auto& input_shape = get_input_partial_shape(0);
 
     OPENVINO_ASSERT(m_variable, "Variable is not initialized.");
-    VariableInfo var_info = {input_shape, element::dynamic, m_variable->get_info().variable_id};
-    NODE_VALIDATION_CHECK(this,
-                          element::Type::merge(var_info.data_type, m_variable->get_info().data_type, arg_t),
-                          "Variables types are inconsistent.");
-    NODE_VALIDATION_CHECK(this,
-                          ov::PartialShape::merge_into(var_info.data_shape, m_variable->get_info().data_shape),
-                          "Variable shape and output shape are inconsistent.");
-    m_variable->update(var_info);
-    set_output_type(0, arg_t, input_shape);
+    const auto& variable_info = m_variable->get_info();
+    const auto& variable_type = variable_info.data_type;
+    const auto& variable_shape = variable_info.data_shape;
+
+    // If no inputs provided, it means this ReadValue doesn't have initial subgraph. This is valid.
+    if (get_input_size() > 0) {
+        const auto& initial_type = get_input_element_type(0);
+        const auto& initial_shape = get_input_partial_shape(0);
+
+        // Variable's shape/type determine a permissible range of values for shape/type inferred from initial_subgraph.
+        // If initial_subgraph is set, then we need to check that shape/type inferred from initial_subgraph
+        // is within the permissible range.
+
+        const auto compatible_type = variable_type.is_dynamic() || initial_type == variable_type;
+        const auto compatible_shape = variable_shape.relaxes(initial_shape);
+
+        OPENVINO_ASSERT(compatible_shape,
+                        "The shape specified in the Variable has to extend (relax) the shape "
+                        "inferred from the initializing subgraph.",
+                        " Variable shape: ",
+                        variable_shape,
+                        " Initialization shape: ",
+                        initial_shape);
+        OPENVINO_ASSERT(compatible_type,
+                        "The type specified in the Variable has to extend (relax) the type "
+                        "inferred from the initializing subgraph.",
+                        " Variable type: ",
+                        variable_type,
+                        " Initialization type: ",
+                        initial_type);
+    }
+
+    set_output_type(0, variable_type, variable_shape);
```

Comment:
@maxnick    Here ReadValue infers its shape and datatype from variable now, instead of input prot 0 or original ngraph node precision (with ppp we can change the node precision here).

Reply:
Yes, we discussed it with @itikhono, so this change is expected.